packages/app/src/heartbeatService.ts

```diff
@@ -90,6 +90,10 @@ export class HeartbeatServiceImpl implements HeartbeatService {
     const date = getUTCDateString();
     if (this._heartbeatsCache?.heartbeats == null) {
       this._heartbeatsCache = await this._heartbeatsCachePromise;
+      // If we failed to construct a heartbeats cache, then return immediately.
+      if (this._heartbeatsCache?.heartbeats == null) {
```

Comment:
@hsubox76 perhaps this should just check if heartbeatsCache === null?

Reply:
I think the root problem is here: https://github.com/firebase/firebase-js-sdk/blob/b782bb2709b661e8b72bde9915352582fb820337/packages/app/src/heartbeatService.ts#L239 or even one function back (readHeartbeatsFromIndexedDB). It's checking to see if whatever comes back from indexedDB is falsy but if it's (as I suspect) an empty object, this will go wrong. We have to make sure that whatever gets returned from this read() function is an object with a heartbeats property which is an array (e.g., matches type `HeartbeatsInIndexedDB`), otherwise we are lying to typescript and so it can't protect us by catching errors further downstream.

Reply:
Ok, I updated the `read()` method (below) to ensure the returned object has a `heartbeats` field.

Reply:
I don't think this check should be needed anymore after the other change but I guess it doesn't hurt to be safe.

Reply:
Yeah, I feel like keeping it place just in case.